test/unit/puppetlabs/puppetserver/lockable_pool_test.clj

```diff
@@ -0,0 +1,191 @@
+                                          (.returnItem pool instance)))
+          lock-thread-started? (promise)
+          lock-acquired? (promise)
+          unlock? (promise)
+          lock-thread (future (deliver lock-thread-started? true)
+                              (.lock pool)
+                              (deliver lock-acquired? true)
+                              @unlock?
+                              (.unlock pool))]
+      @blocked-borrow-thread-started?
+      @lock-thread-started?
+      (is (not (realized? blocked-borrow-thread-borrowed?)))
+      (is (not (realized? lock-acquired?)))
+
+      (return-instances pool instances)
+      (is (not (realized? blocked-borrow-thread-borrowed?)))
+      (is (not (realized? lock-thread)))
+      @lock-acquired?
+
+      (deliver unlock? true)
+      @lock-thread
+      @blocked-borrow-thread-borrowed?
+      @blocked-borrow-thread
+
+      ;; make sure we got here
+      (is (true? true)))))
+
+(deftest pool-lock-reentrant-test
+  (testing "the thread that holds the pool lock may borrow instances while holding the lock"
+    (let [pool (create-populated-pool 2)]
+      (.lock pool)
+      (is (true? true))
+      ;; Current implementation of JRubyPool is not reentrant. This assertion
+      ;; tests that the error we expect is thrown, rather than a deadlock
+      ;; happening. The tests for the behavior we want have been commented out
+      ;; until we change the implementation.
+      (is (thrown? IllegalStateException (.borrowItem pool)))
+      ;; (let [instance (.borrowItem pool)]
+      ;;   (is (true? true))
+      ;;   (.returnItem pool instance))
+
+      (is (true? true))
+      (.unlock pool))))
+
+(deftest pool-lock-reentrant-with-many-borrows-test
+  (testing "the thread that holds the pool lock may borrow instances while holding the lock, even with other borrows queued"
+    (let [pool (create-populated-pool 2)]
+      (.lock pool)
+      (is (true? true))
+      (let [borrow-thread-1 (future (let [instance (.borrowItem pool)]
+                                      (.returnItem pool instance)))
+            borrow-thread-2 (future (let [instance (.borrowItem pool)]
+                                      (.returnItem pool instance)))]
+        ;; this is racey, but the only ways i could think of to make it non-racey
+        ;; depended on knowledge of the implementation
+        (Thread/sleep 500)
+        (is (not (realized? borrow-thread-1)))
+        (is (not (realized? borrow-thread-2)))
+
+        ;; Current implementation of JRubyPool is not reentrant. This assertion
+        ;; tests that the error we expect is thrown, rather than a deadlock
+        ;; happening. The tests for the behavior we want have been commented out
+        ;; until we change the implementation.
+        (is (thrown? IllegalStateException (.borrowItem pool)))
+        ;; (let [instance (.borrowItem pool)]
+        ;;   (is (true? true))
+        ;;   (.returnItem pool instance))
```

Comment:
So, this whole test is effectively commented out because we don't acutually support re-entrant borrows... so it seems like it would probably make sense to just comment out those lines as well?  And worry about race conditions when we move to an implementation that is re-entrant?